File: algorithms/3_hard/F/03336_find_the_number_of_subsequences_with_equal_gcd/dp.c
#define MOD 1000000007

static inline int gcd(int a, int b)
{
    while (b != 0) {
        const int t = b;
        b = a % b;
        a = t;
    }
    return a;
}

int subsequencePairCount(const int* nums, const int len)
{
    assert(len >= 1 && len <= 200 && nums != NULL);
    if (len == 1) {
        return 0;
    }

    int max = 0;
    for (int i = 0; i < len; i += 1) {
        assert(nums[i] >= 1 && nums[i] <= 200);
        const int num = nums[i];
        max = num > max ? num : max;
    }

    const int rows = max + 1;
    const int cols = max + 1;
    int* buff = calloc(2 * (size_t)rows * (size_t)cols, sizeof(*buff));
    assert(buff != NULL);
    int* curr = &buff[0];
    int* next = &buff[rows * cols];
    curr[0] = 1;

    for (int i = 0; i < len; i += 1) {
        const int num = nums[i];

        for (int j = 0; j < rows * cols; j += 1) {
            next[j] = curr[j];
        }

        for (int g1 = 0; g1 <= max; g1 += 1) {
            for(int g2 = 0; g2 <= max; g2 += 1) {
                const int cnt = curr[g1 * cols  + g2];
                if (cnt == 0) { continue; }

                const int idx_g1_pick = gcd(g1, num) * cols + g2;
                const int idx_g2_pick = g1 * cols + gcd(g2, num);

                next[idx_g1_pick] = (next[idx_g1_pick] + cnt) % MOD;
                next[idx_g2_pick] = (next[idx_g2_pick] + cnt) % MOD;
            }
        }

        int* temp = curr;
        curr = next;
        next = temp;
    }

    int res = 0;

    for (int i = 1; i <= max; i += 1) {
        res = (res + curr[i * cols + i ]) % MOD;
    }
/* cleanup: */
    free(buff);
    return res;
}

File: algorithms/3_hard/F/03336_find_the_number_of_subsequences_with_equal_gcd/dp.rs
impl Solution {
    const MOD: i32 = 1_000_000_007;

    #[inline(always)]
    fn gcd(mut a: i32, mut b: i32) -> i32 {
        while b != 0 {
            let t = b;
            b = a % b;
            a = t;
        }
        a
    }

    pub fn subsequence_pair_count(nums: Vec<i32>) -> i32 {
        let len = nums.len();
        assert!(len >= 1 && len <= 200 && nums.iter().all(|num| *num >= 1 && *num <= 200));
        if len == 1 { return 0; }
        let max = *nums.iter().max().unwrap();
        let rows = (max + 1) as usize;
        let cols = (max + 1) as usize;
        let mut buff = vec![0_i32; 2 * rows * cols];
        let (mut curr, mut next) = buff.split_at_mut(rows * cols);
        curr[0] = 1;    

        for num in nums {
            for (i, &cnt) in curr.iter().enumerate() {
                next[i] = curr[i];
            }
            for g1 in 0..=max {
                for g2 in 0..=max {
                    let cnt = curr[(g1 as usize) * cols  + (g2 as usize)];
                    if cnt == 0 { continue; }

                    let idx_g1_pick = (Self::gcd(g1, num) as usize) * cols + (g2 as usize);
                    let idx_g2_pick = (g1 as usize) * cols + (Self::gcd(g2, num) as usize);

                    next[idx_g1_pick] = (next[idx_g1_pick] + cnt) % Self::MOD;
                    next[idx_g2_pick] = (next[idx_g2_pick] + cnt) % Self::MOD;
                }
            }
            (curr, next) = (next, curr);
        }
        
        let mut res = 0_i32;

        for i in 1..=max {
            res = (res + curr[(i as usize) * cols + i as usize]) % Self::MOD;
        }

        res
    }
}
